Add helper to aggregate multiple process groups

pietern opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25768 Add helper to aggregate multiple process groups**

The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

pietern committed:
Add helper to aggregate multiple process groups

The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

mrshenli commented:
Is there an issue requesting this feature? Do we know how much performance gain does this provide.

pietern commented:
@mrshenli I chatted about this with @ngoyal2707 and @myleott. The same technique was used by NVIDIA for their mlperf submission and they say it had a big impact. No need to merge soon by the way; Naman is going to try it out and report back if there are significant gains. Then we can also think a bit more about a nicer public API to actually use this.

pietern committed:
Update on "Add helper to aggregate multiple process groups"


The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

[ghstack-poisoned]

pietern committed:
Update on "Add helper to aggregate multiple process groups"


The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

[ghstack-poisoned]

pietern committed:
Update on "Add helper to aggregate multiple process groups"


The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupRoundRobin.cpp:
```diff
@@ -0,0 +1,102 @@
+namespace c10d {
+
+ProcessGroupRoundRobin::ProcessGroupRoundRobin(
+    std::vector<std::shared_ptr<ProcessGroup>> processGroups)
```
Add checks to enforce same rank and size?

pietern replied:
Added checks.

mrshenli commented on torch/lib/c10d/ProcessGroupRoundRobin.cpp:
```diff
@@ -0,0 +1,102 @@
+  throw std::runtime_error("ProcessGroupRoundRobin does not support recv");
+};
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupRoundRobin::barrier(
```
why can't we support barrier?

pietern replied:
Because we'd need to run a barrier on every process group to ensure all work has completed.

Maybe it is possible after all, but let's address this the moment it is needed.

mrshenli commented on torch/lib/c10d/ProcessGroupRoundRobin.cpp:
```diff
@@ -0,0 +1,102 @@
+};
+
+const std::shared_ptr<ProcessGroup>& ProcessGroupRoundRobin::next() {
+  auto& processGroup = *iterator_;
```
do we need lock?

pietern replied:
No lock needed. All process groups can only be used by a single thread anyway (because the ordering of the calls must be globally deterministic).

mrshenli replied:
> All process groups can only be used by a single thread anyway

Does this still hold if we use tags? IIUC, we do support using ProcessGroup with different tag in different thread, do we want the same thing for ProcessGroupRoundRobin as well? If no, let's explicitly say that in the comments/docs.

pietern replied:
We only support that by accident, for send/recv only on ProcessGroupGloo. In general, collective operations must be issued in globally deterministic order, and any races here would immediately lead to incorrectness. Adding a mutex here would give a false sense of safety, where in reality there is none.

For example, if you have two threads T1 and T2, and both want to execute an allreduce. T1 operates on 1 float and T2 operates on 10 floats. Now consider two processes that do this. There is nothing we can do to ensure the correct ordering of these operations if the application code doesn't already take care of it. Without the application doing anything, you could easily end up with T1 trying to work with T2 on the other process, and causing a crash/

pietern replied:
Just checked and the support in ProcessGroupGloo is accidental. There are no facilities for thread safety there at all. That it works in the first place is just because send/recv pretty much just patching through to Gloo directly.

mrshenli reviewed (commented):
As this is not yet a public API for now, I guess it is fine to have no doc and limited test coverage. Any plan to make it a public API?

Test failure is real:

```
Oct 29 08:57:22 ======================================================================
Oct 29 08:57:22 FAIL: test_round_robin (__main__.ProcessGroupGlooTest)
Oct 29 08:57:22 ----------------------------------------------------------------------
Oct 29 08:57:22 Traceback (most recent call last):
Oct 29 08:57:22   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 130, in wrapper
Oct 29 08:57:22     self._join_processes(fn)
Oct 29 08:57:22   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 211, in _join_processes
Oct 29 08:57:22     self._check_return_codes(elapsed_time)
Oct 29 08:57:22   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 235, in _check_return_codes
Oct 29 08:57:22     self.assertEqual(first_process.exitcode, 0)
Oct 29 08:57:22   File "/var/lib/jenkins/workspace/test/common_utils.py", line 767, in assertEqual
Oct 29 08:57:22     super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
Oct 29 08:57:22 AssertionError: 10 not less than or equal to 1e-05 : 
```

pietern committed:
Update on "Add helper to aggregate multiple process groups"


The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

[ghstack-poisoned]

pietern commented:
Fixed the broken test and addressed the comments.

pietern committed:
Update on "Add helper to aggregate multiple process groups"


The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

[ghstack-poisoned]

pietern committed:
Update on "Add helper to aggregate multiple process groups"


The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

[ghstack-poisoned]

zhaojuanmao commented on test/test_c10d.py:
```diff
@@ -1509,6 +1509,23 @@ def test_barrier_implies_wait(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        pg = c10d._round_robin_process_groups([
+            c10d.ProcessGroupGloo(
+                c10d.PrefixStore(str(i), store),
```
will it be good to add comments somewhere to suggest users to use PrefixStore in sub process group?

pietern replied:
Good point.

This should be part of the general docs about the lower level c10d API. It's somewhat implied, since a single store being used by multiple process groups intuitively gets clobbered, but agree it would be good to include in docs somewhere.

pietern commented:
The test failures are unrelated.

@mrshenli Can you take another look at this?

mrshenli reviewed (commented):
This looks great to me. I only have two minor concerns on thread-safety and test coverage. I would assume this is used in elastic training? In so, should we add a test for destroy and recreate a round robin PG?

pietern committed:
Update on "Add helper to aggregate multiple process groups"


The round robin process group can be constructed from multiple other
process groups. Every collective call against this new process group
is delegated to the specified process groups in a round robin fashion.

Doing so may benefit performance when calling into multiple NCCL
process groups. Instead of adding support for round-robin usage of
NCCL communicators, we achieve the same without changing the NCCL
process group and adding this wrapper class.

The API to create this round robin process group is a bit harsh. If we
find it adds significant benefit we can revisit and make this a first
class citizen in the torch.distributed module.

Differential Revision: [D17226323](https://our.internmc.facebook.com/intern/diff/D17226323/)

[ghstack-poisoned]

pietern commented:
@mrshenli Thanks. I added a test to exercise create/use/destroy multiple times.

pietern commented:
Failures are unrelated.

mrshenli reviewed (approved):
LGTM! Thanks!
